### PR TITLE
Add ability to watch diagram and csx files for changes

### LINKED ISSUE #265: feature - `ss.cli watch` automated code generation when diagram saved
From https://github.com/StateSmith/StateSmith/issues/264:

> In StateSmith today, saving the image file and generating the code are two separate operations. The user must first save, and then run a csx file to generate the code. The tool does not necessarily know which config to use, so it cannot generate the files.

> With zero-config, the second step will no longer be necessary. Because most models will not need a config file, the drawio plugin can simple run the default codegen automatically after every save, thereby keeping the generated files in sync with the svg on every save.

I really like this idea, but it will take some time to get to "zero-config".

A very nice interim step could be to add "watch" functionality to the StateSmith.Cli tool. Something kind of like the watch command for typescript. Watch these diagram/csx files and run code generation when they change.

Could be something like `ss.cli watch` and use the statesmith.cli.json manifest file or work recursively in the current directory.

Might not even be that hard as ss.cli already does a good chunk of that work.

### PR DESCRIPTION
**Pull request by emmby**
Usage: add -w to any run to keep the tool open and watch for changes.
eg. StateSmith.Cli run --lang=JavaScript -hr -w
StateSmith.Cli run --help for more info.

Limitations:
- currently reprocesses every file in the scan if any of the files change. Ideally it would process just the changed file, but that requires refactoring ScanResults in a more invasive way.

Closes #265

### CODE REVIEW

**Comment by adamfk** [comment]
Very cool! Will try out in a few days when I'm back in action. I'd like to follow up (in a new PR or this one) to process just the changed file.

Thanks!

**Comment by adamfk** [comment]
Thoughts on what should happen if code gen fails? Maybe from invalid diagram? 

Thoughts on whether we should someday have an option to discover diagrams created after watch was started? I think typescript does this.

**Comment by emmby** [comment]
Good point on the invalid diagram. It should probably continue watching for changes. I'll revise.

Open to a future watch option to discover diagrams, that could make sense. Probably the way to do it would be to allow the user to pass in directories in the file list, and watch those directories for new files that match the supported file extensions.

File: src/StateSmith.Cli/Run/RunHandler.cs [hunk 1]
```diff
@@ -98,6 +99,37 @@ private void RunInner()
...
+                var path = Path.GetDirectoryName(diagramFile);
+                watcher.Path = path!=null && path.Length>0 ? path : "."; 
+                watcher.Filter = Path.GetFileName(diagramFile);
+                _runConsole.WriteLine($"Watching {diagramFile}");
+                _runConsole.WriteLine($"Path: {watcher.Path}");
+                _runConsole.WriteLine($"Filter: {watcher.Filter}");
+                watcher.Changed += (sender, e) => 
+                {
+                    _runConsole.WriteLine($"File {diagramFile} has changed.");
+                    // TODO only process the changed file. Not sure how to break up ScanResults to do this
```
Comment (adamfk) [inline src/StateSmith.Cli/Run/RunHandler.cs:120]: if we just poll, or inspect the file extension, we could call the relevant handler [for .csx](https://github.com/StateSmith/StateSmith/blob/7206f7e387f09eefcaf4847609bc55f34d261a28/src/StateSmith.Cli/Run/RunHandler.cs#L232) and we could refactor this [bit of code for diagrams](https://github.com/StateSmith/StateSmith/blob/7206f7e387f09eefcaf4847609bc55f34d261a28/src/StateSmith.Cli/Run/DiagramRunner.cs#L86-L111) into a public method.

Note that the scan will pick up diagrams that are processed by .csx files. Here's where this check is currently done: https://github.com/StateSmith/StateSmith/blob/7206f7e387f09eefcaf4847609bc55f34d261a28/src/StateSmith.Cli/Run/DiagramRunner.cs#L49-L59

**Comment by adamfk** [comment]
@emmby I'm excited to start using the watch feature. It's a great idea! Do you have further plans for this PR, or would you mind if I merge and continue working on it to overcome the above limitations?

**Comment by emmby** [comment]
Sorry, I've been preoccupied with a laundry room remodel. Please feel free to merge the pr and/or evolve it

**Comment by adamfk** [comment]
No worries! I think I just found a way to add another algorithm to
StateSmith easy so I'll probably focus on that first.

**Comment by adamfk** [comment]
Hi Mike, this PR ended up being rather behind main and I know you're busy so I copied some of the code and implemented over here: https://github.com/StateSmith/StateSmith/pull/413

Thanks for the initial concept! I think the feature will evolve a bit still, but it is already really nice :)

**Pull request closed without merging**